src/compiler/crystal/semantic/method_lookup.cr

```diff
@@ -250,6 +250,12 @@ module Crystal
 
         match_arg_type = arg_type.restrict(arg, context)
         if match_arg_type
+          if !named_args && !splat_arg_types && match_arg_type.same?(arg_type) && arg_types.size == 1
+            # Optimization: no need to create matched_arg_types if
+            # the call has a single argument and it exactly matches the restriction
+            break
+          end
```

Comment:
Far down in the method we return `matched_arg_types || arg_types`. That is, if we end up with no `matched_arg_types` we use `arg_types`. The optimization here is avoiding creating `matched_arg_types` with exactly one type that's going to be exactly the same as `arg_types`.

Believe it or not, calls with just one argument and matching this condition is super common! In the compiler it happens a bit more than 200K times. So that's 200K less arrays we allocate.